Support the free-threaded build of Python

I'm working full-time with a team @quansight-labs on ecosystem support for free-threaded Python and am happy to help out here. I'm also one of the main authors of https://py-free-threading.github.io which might also answer some questions you might have.

If I clone the repo and run the tox tests on 3.13t and 3.14t, all the tests pass with the GIL disabled. I see that 3.13t and 3.14t are already in your testing matrix.

If I run the tests under [`pytest-run-parallel`](https://github.com/Quansight-Labs/pytest-run-parallel), which executes each test many times in a per-test thread pool, I do see one failure:

```
________________ ERROR at call of test_default_ref_counting _________________

    @pytest.mark.skipif(
        hasattr(sys, "pypy_version_info"), reason="PyPy uses incompatible GC"
    )
    def test_default_ref_counting():
        class DefaultRefCountingClass:
            def __init__(self, value):
                self._value = value

            def convert(self):
                if self._value > 1:
                    return type(self)(self._value - 1)
                return 0

        import gc

        gc.collect()
        ujson.dumps(DefaultRefCountingClass(3), default=lambda x: x.convert())
>       assert not any(
            type(o).__name__ == "DefaultRefCountingClass" for o in gc.get_objects()
        )
E       assert not True
E        +  where True = any(<generator object test_default_ref_counting.<locals>.<genexpr> at 0x30622140620>)

tests/test_ujson.py:1097: AssertionError
```

But that's not surprising because this test relies on global state in the GC and isn't thread-safe.

I'm not sure what the thread safety story of the ultrajson C library is. I looked over the Python bindings and didn't see any obvious issues (e.g. C global state).

It'd probably be worthwhile to write some explicitly multithreaded tests of some multithreaded workflows - e.g. parsing many json files in a thread pool - and run the tests under TSAN. See https://py-free-threading.github.io/thread_sanitizer/ for more detail.

--- a/.github/workflows/test.yml
+++ b/.github/workflows/test.yml
@@ -43,11 +43,6 @@ jobs:
           cache: pip
           cache-dependency-path: "setup.py"
 
-      - name: Set PYTHON_GIL
-        if: endsWith(matrix.python-version, 't')
-        run: |
-          echo "PYTHON_GIL=0" >> "$GITHUB_ENV"
-
       - name: Install dependencies
         run: |
           python -m pip install -U pip

--- a/src/ujson/python/ujson.c
+++ b/src/ujson/python/ujson.c
@@ -200,5 +200,8 @@ PyMODINIT_FUNC PyInit_ujson(void)
     return NULL;
   }
 
+#ifdef Py_GIL_DISABLED
+  PyUnstable_Module_SetGIL(module, Py_MOD_GIL_NOT_USED);
+#endif
   return module;
 }

--- a/tests/test_ujson.py
+++ b/tests/test_ujson.py
@@ -1683,6 +1683,15 @@ def test_comprehensive_json_fixture():
     assert ujson.loads("1.") == 1.0
 
 
+def test_gil_not_reenabled():
+    try:
+        gil_enabled = sys._is_gil_enabled()
+        gil_expected = bool(sys.flags.gil)
+    except AttributeError:
+        pytest.skip(reason="Free threading properties unavailable")
+    assert gil_enabled is gil_expected
+
+
 """
 The following checks are not part of the standard test suite.
 They can be run manually as follows:

